### ISSUE TITLE
Constructor matcher downvotes arguments with defaults, which may choose least correct option.

### ISSUE THREAD

**Issue by BlueTufa** [json4s/json4s#661]
### json4s version
3.6.7 + 
### scala version
2.12.11
### jdk version
8.0

Given the following scenario:
```case class BadSpec(item2: Int, item3: Int, isVisited: Boolean = false)
case object BadSpec {
  def apply(item1: Int, item2: Int, item3: Int): BadSpec = BadSpec(item2, item3)
}        
 serialization.read[BadSpec](s"""{"item2": 789, "item3": 123}""") must_== BadSpec(789, 123)
```
The constructor matcher will downvote the isVisited arg even though there is a default, leading to the least correct ctor being selected.

**Comment by mccartney** [comment]
Also - the `else` line looks fishy to me:
https://github.com/json4s/json4s/blob/master/core/src/main/scala/org/json4s/reflect/descriptors.scala#L60

Is it really meant to give absolute score of `-100` for the whole constructor or for that argument only. I am debugging (what turned out to be the original issue here) and all my constructors pretty much score `-99` regardless of argument count, etc, as they get `1` for the first one and `-99` for some of the non-matching ones later in the tail.

In other words. Shouldn't it be: `s-100` instead of `s`?

**Comment by BlueTufa** [comment]
It's actually tracking the score for the entire constructor.  My understanding is the following:
1.  If the ctor argument name does not match the key name of the incoming json, AND
2.  The argument is not optional, AND
3.  The argument does not have a default value -- *this was my proposed bug fix*, THEN
4.  Return -100 as the score.

I think you're right, `s-100` would be likely lead to a better score.  However, if items #2 and #3 above are false, then the likelihood of a runtime error is very high.  If you have a code snippet of your scenario where everything got voted -99, I would be very curious to see it.  

Thanks!

**Comment by mccartney** [comment]
I think your fix in #662 is right.

I would need to spend time extracting to a minimal example without all these dependencies, but based on what I see now, the scenario I am debugging (without your #662 fix) could likely be modeled in a unit test with:
```
{first: "John", last: "Smith", age: 48, city: "Kosice"}
```
```
case class Person {
  first: String,
  last: String,
  age: Int,
  city: String,
  country: String = "Slovakia",
  height: Int = 180,
  siblings: Int = 0,
  parents: Int = 2
}
object Person {
  def apply(somethingTotallyDifferent: TotallyUnrelatedType, last: String) = ...
  def apply(first: String, last: String, permissions: List[Permission]) = ...
}
```

So we have three constructors:
- one from the case class - which should be picked, but gets `-99`
- the one with `TotallyUnrelatedType` - which gets picked by `json4s:3.6.7` and gets `-99`
- the one with `List[Permissions]` which should not get picked, it also gets `-99`

Disclaimer: I am not sure if this is a representative scenario and not sure of the actual scores, because the real one I am having is deeper down in the object hierarchy (and more complex) and I am not sure if this nesting does not play a role here.

**Comment by mccartney** [comment]
I don't see any tests for `bestMatching` in https://github.com/json4s/json4s/blob/master/tests/src/test/scala/org/json4s/ReflectorSpec.scala

**Comment by BlueTufa** [comment]
Your sample makes sense.  It is very similar to one of the bugs we were seeing after the upgrade to Json4s 3.6.7.  

Your case class ctor is the best match, but it's being downvoted due to the fact that arguments with default values are not being weighted properly.  An excellent example of the bug.  

And of course as a workaround you could create a companion ctor with only the required arguments.  For us it proved to be impractical to create many additional companion ctors.

Thank you
